### PR TITLE
Cvar manager

### LINKED ISSUE #8: Cvar manager
## Description

Create manager for cvar.
We should know what plugin create a cvar.
Cvar access from plugin should be faster than now, we don't have any cache system.

- Operating system:

- ReHLDS version:

- Metamod version:

- SPMod version:

## Other Comments
Issue for projects list

### LINKED ISSUE #11: Convert cvar API to methodmap
## Description
Depends on #8.
Natives which access cvar by name could be entirely dropped in favor of pcvars.

Possible api:
```sourcepawn
enum Cvar
{
    INVALID_CVAR = -1
};

methodmap Cvar
{
    /* this constructor would register cvar */
    public native Cvar(const char[] name, const char[] value, CvarFlags = None);
    public native float GetFloat();
    public native void GetString(char[] value, int size);
    public native int GetInt();
    public native CvarFlags GetFlags();
    public native void SetFloat(float value);
    public native void SetString(const char[] value);
    public native void SetInt(int value);
    public native void SetFlags(CvarFlags flags);
};
native Cvar FindCvar(const char[] name);
```

- Operating system: n/a
- ReHLDS version: n/a
- Metamod version: n/a
- SPMod version: n/a

## How do we replicate the issue?
n/a

## Expected behavior (i.e. solution)
Methodmap API to work with cvar.

## Other Comments
n/a

### PR DESCRIPTION
**Pull request by Garey27**
## Description

Implementation of Cvar manager
- Cvar class
- Cvar manager class
- Interface to cvar and cvarmanager
- Cvar caching
- Cvar callbacks
- Methodmap for cvar

## Motivation and Context

Fix #8
Fix #11 

## How has this been tested?

```sourcepawn
public void TestCvarChange_Callback(Cvar cvar, const char[] oldValue, const char[] newValue)
{
        printToServer("Cvar changed:%s\n", newValue);
}
public void OnPluginInit()
{
	printToServer("Create cvar with value = 5.0\n");
	Cvar c = Cvar("test", "5.0", None);
	printToServer("Add callback to cvar\n");
	c.AddHookOnChange(TestCvarChange_Callback);
	printToServer("Print as float\n");
	printToServer("%f\n", c.GetFloat());
	printToServer("Set value to 2.0\n");
	c.SetFloat(2.0);
	printToServer("Print as float\n");
	printToServer("%f\n", c.GetFloat());
	printToServer("Set value to 1337\n");
	c.SetInt(1337);
	printToServer("Print as int\n");
	printToServer("%d\n", c.GetInt());
}
```
## Screenshots (if appropriate):
photo_2018-06-19_08-24-43
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Garey27** [comment]
Need another review 🐮

**Comment by Garey27** [comment]
Sorry, no need review atm, need to fix (I wasn't opened you request full list and fixed only part lol)

File: src/CvarSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        m_cvars.emplace(name, cvar);
+        // Raise cvar num
+        m_id++;
+        return cvar.get();
+    }
+    // Not found
+    return nullptr;
+}
+
+std::shared_ptr<Cvar> CvarMngr::findCvarCore(std::string_view name)
```
Comment (Amaroq7) [inline src/CvarSystem.cpp]: ```findCvarCore(std::string_view name)``` should look like
```cpp
std::shared_ptr<Cvar> CvarMngr::findCvarCore(std::string_view name, bool cacheonly)
{
    auto pair = m_cvars.find(name.data());
    if (pair != m_cvars.end())
        return pair->second;

    if (cacheonly)
        return nullptr;

    cvar_t *pcvar = nullptr;
    pcvar = CVAR_GET_POINTER(name);
    if (pcvar)
    {
        std::shared_ptr<Cvar> cvar = std::make_shared<Cvar>(name, m_id, pcvar->string, static_cast<ICvar::Flags>(pcvar->flags), pcvar);
        // Always add to cache
        m_cvars.emplace(name, cvar);
        // Raise cvar num
        m_id++;
        return cvar.get();
    }
    // Not found
    return nullptr;
}
```
and ```findCvar```
```cpp
ICvar *CvarMngr::findCvar(const char *name)
{
    return findCvarCore(name, false).get();
}
```

File: src/RehldsApi.cpp [hunk 2]
```diff
@@ -42,40 +42,14 @@ static void SV_DropClientHook(IRehldsHook_SV_DropClient *chain,
...
-    if (!strcmp(cvar->string, value))
-    {
-        chain->callNext(cvar, value);
-        return;
-    }
-
-    std::shared_ptr<Forward> forward = gSPGlobal->getForwardManagerCore()->getDefaultForward(def::CvarChange);
-    if (!forward)
+{   
+    auto cachedCvar = gSPGlobal->getCvarManagerCore()->findCvarCore(cvar->name);
```
Comment (Amaroq7) [inline src/RehldsApi.cpp]: According to my other comment.
```cpp
auto cachedCvar = gSPGlobal->getCvarManagerCore()->findCvarCore(cvar->name, true);
```

File: src/CvarsNatives.cpp [hunk 3]
```diff
@@ -19,147 +19,316 @@
...
+    enum { arg_name = 1 };
+    const std::unique_ptr<CvarMngr> &cvarMngr = gSPGlobal->getCvarManagerCore();
+    char *cvarName;
+    ctx->LocalToString(params[arg_name], &cvarName);
 
-    static cvar_t cvarToRegister;
-    cvarToRegister.name = cvarName;
-    cvarToRegister.string = cvarValue;
-    cvarToRegister.flags = params[3];
+    auto plCvar = cvarMngr->findCvarCore(cvarName);
```
Comment (Amaroq7) [inline src/CvarsNatives.cpp]: According to my other comment.
```cpp
auto plCvar = cvarMngr->findCvarCore(cvarName, false);
```

File: src/CvarsNatives.cpp [hunk 4]
```diff
@@ -19,147 +19,316 @@
...
+    const std::unique_ptr<CvarMngr> &cvarMngr = gSPGlobal->getCvarManagerCore();
+    auto cvar = cvarMngr->findCvarCore(cvarId);
     if (!cvar)
+    {
+        ctx->ReportError("Cvar not found!");
         return 0;
+    }
 
-    return cvar->flags;
+    return (cell_t)cvar->getFlags();
```
Comment (Amaroq7) [inline src/CvarsNatives.cpp]: I don't think cast is needed here.

File: src/CvarSystem.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                              ICvar::Flags flags)
+{
+    return registerCvarCore(name, value, flags).get();
+}
+
+std::shared_ptr<Cvar> CvarMngr::registerCvarCore(std::string_view name,
+                                                 std::string_view value,
+                                                 ICvar::Flags flags)
+{
+    if (auto cvar = findCvarCore(name))
```
Comment (Amaroq7) [inline src/CvarSystem.cpp]:
```cpp
auto cvar = findCvarCore(name, true)
```

**Comment by Amaroq7** [comment]
Looks good to me. Thanks for work done implementing it.

**Pull request merged**